### PR TITLE
drivers/flash/flash_simulator: Allow unaligned read

### PR DESCRIPTION
**Pull request by rakons**
This commit adds the default configuration to the flash_simualtor
to allow unaligned reads.

This should fix the #16275 issue - I mean, I have tested in the other program, and Settings works as expected with FCB backend. I have not tested in exactly the same situation as in the listed issue.

### CODE REVIEW

**Comment by rakons** [comment]
CC @lemrey @kapi-no @nvlsianpu 
@ru-fu - Could you check the documentation, please?

File: drivers/flash/flash_simulator.c
Comment (nvlsianpu) [inline drivers/flash/flash_simulator.c]: use IS_ENABLED() macro

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_simulator.c
Comment (kapi-no) [inline drivers/flash/flash_simulator.c]: Consider moving CONFIG_FLASH_SIMULATOR_UNALIGNED_READ to the outer if statement. This way, the feature logic is more pronounced and easier to follow.
```suggestion
	if (!CONFIG_FLASH_SIMULATOR_UNALIGNED_READ) {
	(...)
	}
```

Comment (rakons) [reply drivers/flash/flash_simulator.c]: Fixed.

Comment (rakons) [reply drivers/flash/flash_simulator.c]: I have merged your proposal and @nvlsianpu .

**Review by kapi-no — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

**Comment by nvlsianpu** [comment]
cc @lemrey

**Review by lemrey — APPROVED** [review]
Thank you @rakons .

File: drivers/flash/Kconfig.simulator
Comment (ru-fu) [inline drivers/flash/Kconfig.simulator]: If selected, the reading operation does not check if access is aligned.
         Enable this option only if you want to simulate a specific FLASH interface
	 that requires alignment read access.

Comment (rakons) [reply drivers/flash/Kconfig.simulator]: I just saw that there is a logical mistake, should be:
**Disable** this option only...

Thank you. I will update it after we get shippable green.

Comment (ru-fu) [reply drivers/flash/Kconfig.simulator]: That suddenly makes a lot more sense. ;)

Comment (rakons) [reply drivers/flash/Kconfig.simulator]: Done.

**Review by ru-fu — APPROVED** [review]

**Comment by rakons** [comment]
The failing here https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/42516/5/console, seems not to be related to my work, is it?

**Comment by nvlsianpu** [comment]
@carlescufi can you merge this- shippable failure is not related to this patch content at all.

File: drivers/flash/Kconfig.simulator
Comment (ioannisg) [inline drivers/flash/Kconfig.simulator]: "aligned read access"
"or alignment when performing read access"

File: tests/drivers/flash_simulator/prj.conf
Comment (ioannisg) [inline tests/drivers/flash_simulator/prj.conf]: This change is not reflected in the commit message.

**Comment by carlescufi** [comment]
Well, I'd rather not merge until we know the cause of the shippable failure and fix it. Can you please rebase on top of the latest master to see if it's fixed there?

File: tests/drivers/flash_simulator/prj.conf
Comment (ioannisg) [reply tests/drivers/flash_simulator/prj.conf]: (Would be nice to justify why the test requires aligned access)

Comment (rakons) [reply tests/drivers/flash_simulator/prj.conf]: This is just because test checks this part and I do not wish to implement 2 different tests (repeating the whole test with just this parts disabled does not add goods).
I am not sure if we should explain everything in the PR message? PR message should be short not coming into details? Anyone? Should I really describe in PR message that I have added this flag and why?

Comment (ioannisg) [reply tests/drivers/flash_simulator/prj.conf]: The introduction of this Kconfig flag is well-described in the commit message. Add a sentence explaining why the tests needs to run with the Kconfig symbol undefined.

File: drivers/flash/Kconfig.simulator
Comment (rakons) [reply drivers/flash/Kconfig.simulator]: Fixed to "aligned read access".

File: tests/drivers/flash_simulator/prj.conf
Comment (rakons) [reply tests/drivers/flash_simulator/prj.conf]: A note added.

**Comment by Laczen** [comment]
@nvlsianpu , @carlescufi , this fix is hiding a problem in settings/FCB. With all the code that is added to settings to get the alignment good, this error should not appear. The fix also means that the settings/FCB combination can only be used on system with unaligned.read possibility.

**Comment by rakons** [comment]
@Laczen, this PR does not hide anything. Most of the available FLASH memories requires aligned write access and allows reading even a single byte. I have checked a few implemented FLASH drivers and every reading function in flash driver layer I saw uses memcpy in the bottom, so it has to support unaligned access by design. We can eventually document this fact.

Even look at the names of the constants used here: ```FLASH_SIMULATOR_PROG_UNIT``` - programming unit, not reading unit.

I can think only about 2 kinds of memories that do not allow free byte read: Some specific SPI memories where access is always 16 bit and NAND flash with sequential access by its nature (I am considering an SSD disks and memory cards to be NAND flash with a controller, so please do not be nitpicking about this).

FCB always writes data with alignment required by the FLASH backend. And reads only the number of bytes it requires. As a matter of fact, the read address here is in required **WRITE** boundary, but the number of bytes to read is smaller.

**Comment by Laczen** [comment]
@rakons, if you look at the settings implementation a lot of code is added to get the alignment good. If a unaligned read would be possible this code would not be required. However it seems that even with all the added code there still is a requirement to do an unaligned read (otherwise your fix would not be required). So there is something wrong in the code that is hidden by providing unaligned read.
@rakons , I am not saying your fix is not good, most devices will allow a unaligned read.

**Comment by rakons** [comment]
@Laczen, please create another issue and start the discussion there.

**Comment by Laczen** [comment]
@rakons, maybe something else is wrong and the flash simulator sets the alignment requirements wrong. Your proposal fixes the problem, but it misses the root cause. Do you really want me to recreate the same problem: #16275 fails with `CONFIG_FLASH_SIMULATOR_UNALIGNED_READ=n`.

**Comment by nvlsianpu** [comment]
@Laczen @rakons @carlescufi I remember that some time ago we were disusing the same problem - as I remember almost all driver we have in zephyr supports unaligned read. Only one driver (will find the name) support read at aligned offset but sill have had no requirements regards read-out length. FCB implementation ensure that each FCB field begins at aligned offset. Problem for flash-simulator was that it emulates different behavior (assert aligned offset and read-out size).

I think that each flash driver should support unaligned access - I will add this to API meetings topics list.

**Comment by nvlsianpu** [comment]
@Laczen What is your opinion on flash driver read API?

**Comment by Laczen** [comment]
@nvlsianpu, my opinion on the flash driver read api is that it should provide support to read at least starting from a aligned position and read an unaligned length. A further enhancement is to provide a starting from a unaligned position (this could be optional but nice).
@nvlsianpu, the code in the settings api is written in such a way that the reads are aligned for start and length, so it should not trigger an assert.

**Comment by rakons** [comment]
No it is not. Line 32 in fcb_elem_info.c:
```
rc = fcb_flash_read(fcb, loc->fe_sector, loc->fe_elem_off, tmp_str, 2);
```
The position is aligned. The size is not.
Line 79 in the same file:
```
u8_t fl_crc8;
...
rc = fcb_flash_read(fcb, loc->fe_sector, off, &fl_crc8, sizeof(fl_crc8));
```

The fcb.c file, line 266:
```
rc = fcb_flash_read(fcb, sector, 0, fdap, sizeof(*fdap));
```
Sure, it is aligned to 4.... or 2 if we use 8-16 bit CPU.... or 8 if we use 64 bit CPU... but there is no control about if it is aligned to ```FLASH_SIMULATOR_PROG_UNIT```.

I do not even want to search more. If memory requires aligned read access this can always be fixed in the driver layer to allow to read unaligned data. The write is problematic. When reading you can always throw away some data.

**Comment by carlescufi** [comment]
Merging, since the alignment issues are unrelated to this PR and will be tracked by an issue raised by @nvlsianpu

**Comment by Laczen** [comment]
@rakons, now you have found the root cause: **fcb requires a unaligned read** (but probably only a unaligned length). @nvlsianpu, this means that the read length alignment that is done in settings could be simplified.
@rakons, to make the flash_simulator capable of simulating memories that have requirements on the alignment it might be good to create 2 defines: `CONFIG_FLASH_SIMULATOR_ALIGNED_READ_START` and `CONFIG_FLASH_SIMULATOR_ALIGNED_READ_LENGTH`, with the second automatically selecting the first.

**Comment by nvlsianpu** [comment]
discussion moved to #16438

**Comment by rakons** [comment]
@Laczen 
> alignment it might be good to create 2 defines: `CONFIG_FLASH_SIMULATOR_ALIGNED_READ_START` and `CONFIG_FLASH_SIMULATOR_ALIGNED_READ_LENGTH`, with the second automatically selecting the first.

This is something I treat as valuable input inside the PR discussion - highly related to the PR scope and improves the proposed solution.

**Comment by Laczen** [comment]
@rakons, to get to this point the discussion was required. I know that inside the settings module (not fcb code) there is code that supposes that a flash interface with aligned read and length should be supported, so I guessed (wrongly) that fcb supported this. The code inside the settings module is not required because fcb by itself does not support a flash interface with aligned read and length.

I am not here to nag about something. I saw the fix and asked myself why does this fix something that should not be there in the first place (as the settings module does all the work to support aligned read and length, I supposed that fcb did this as well). This is why I said that the PR was hiding something.

It's a pity that your PR was already merged.

**Comment by rakons** [comment]
@Laczen, sadly us cannot know a destined. Issue is closed. Keep this idea by creating an issue for it or describing it in the one already created.

**Comment by rakons** [comment]
@Laczen - I have even created us the issue: #16444. Could we continue the future discussion there?

**Pull request merged**
